Support `vendor.yaml` as a new way of vendoring components, stacks or any other artifacts (#448)
cloudposse/atmos: website/docs/core-concepts/vendoring/vendoring.md

```bash
atmos vendor pull -c <component>
```

<br/>

:::tip
Refer to [`atmos vendor pull`](/cli/commands/vendor/pull) CLI command for more details
:::

## Vendoring Manifest

To vendor remote artifacts, create a `vendor.yaml` file similar to the example below:

```yaml
apiVersion: atmos/v1
kind: AtmosVendorConfig
metadata:
  name: example-vendor-config
  description: Atmos vendoring manifest
spec:
  # `imports` or `sources` (or both) must be defined in a vendoring manifest
  imports:
    - "vendor/vendor2.yaml"
    - "vendor/vendor3.yaml"

  sources:
    # `source` supports the following protocols: OCI (https://opencontainers.org), Git, Mercurial, HTTP, HTTPS, Amazon S3, Google GCP,
    # and all URL and archive formats as described in https://github.com/hashicorp/go-getter.
    # In 'source', Golang templates are supported  https://pkg.go.dev/text/template.
    # If 'version' is provided, '{{.Version}}' will be replaced with the 'version' value before pulling the files from 'source'.
    # Download the component from the AWS public ECR registry (https://docs.aws.amazon.com/AmazonECR/latest/public/public-registries.html).

>     - component: "vpc"

Reviewer: Is there anyway to vendor a component and copy a local file over a vendored file? Such as if I'm using my own `providers.tf` mixin and I want to continue to vendor components without manually overwriting the mixin using `cp` ?

Reviewer 2: Not in a DRY way. The best you can do is `exclude` the `providers.tf` so it won't overwrite your local one.

Reviewer: nvm, this is already implemented using `mixins`. thank you.

Reviewer 2: Maybe we could consider a `post_pull_hook` or something like that can run a command. Then you can do the `cp` or anything else. Dont' think we want to support a specialized case around vendoring and then copying another file.

Reviewer 2: You can have another `source` and `target` defined to vendor it from a local file, but that will be ugly (non-DRY)

Reviewer: You mean something like this

```yaml
    - component: "vpc-flow-logs-bucket"
      source: "github.com/cloudposse/terraform-aws-components.git//modules/vpc-flow-logs-bucket?ref={{.Version}}"
      version: "1.323.0"
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}"
    - component: "vpc-flow-logs-bucket-mixins"
      # if in components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}/
      source: "../../../../../mixins/providers.tf"
      version: "1.323.0"
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}"
```

It would be nice if `source` then was also a list, no?

```yaml
    - component: "vpc-flow-logs-bucket"
      source:
        - "github.com/cloudposse/terraform-aws-components.git//modules/vpc-flow-logs-bucket?ref={{.Version}}"
        - "../../../../../mixins/providers.tf"
      version: "1.323.0"
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}"
```

      targets:
        - "components/terraform/infra/vpc3"
      # Only include the files that match the 'included_paths' patterns.
      # If 'included_paths' is not specified, all files will be matched except those that match the patterns from 'excluded_paths'.
      # 'included_paths' support POSIX-style Globs for file names/paths (double-star `**` is supported).
      # https://en.wikipedia.org/wiki/Glob_(programming)
      # https://github.com/bmatcuk/doublestar#patterns
      included_paths:
        - "**/*.tf"
        - "**/*.tfvars"
        - "**/*.md"
    - component: "vpc-flow-logs-bucket"
      source: "github.com/cloudposse/terraform-aws-components.git//modules/vpc-flow-logs-bucket?ref={{.Version}}"
      version: "1.323.0"
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}"
      excluded_paths:
        - "**/*.yaml"
        - "**/*.yml"
```

<br/>

- The `vendor.yaml` vendoring manifest supports Kubernetes-style YAML config to describe vendoring configuration for components, stacks,
  and other artifacts. The file is placed into the directory from which the `atmos vendor pull` command is executed (usually the root of the repo)

> - The `sources` in `vendor.yaml` support all protocols (local files, Git, Mercurial, HTTP, HTTPS, Amazon S3, Google GCP), and all URL and
>   archive formats as described in [go-getter](https://github.com/hashicorp/go-getter), and also the `oci://` scheme to download artifacts from
>   [OCI registries](https://opencontainers.org).

Reviewer 2: ```suggestion
- The `sources` in `vendor.yaml` support all protocols (local files, Git, Mercurial, HTTP, HTTPS, Amazon S3, Google GCP), and all URL and
  archive formats as described in [go-getter](https://github.com/hashicorp/go-getter), and also the `oci://` scheme to download artifacts from
  [OCI registries](https://opencontainers.org). **IMPORTANT** include the `{{ .Version }}` parameter in your `source` URI to ensure the correct version of the artifact is downloaded. For example,
  ```source: "github.com/cloudposse/terraform-aws-components.git//modules/vpc-flow-logs-bucket?ref={{.Version}}"
  ```
```

> - The `targets` in the sources support absolute and relative paths (relative to the directory where the command is executed)

Reviewer 2: ```suggestion
- The `targets` in the sources support absolute and relative paths (relative to the directory where the command is executed). Multiple targets can be specified.
 
  Here's an example of how multiple targets might be used to ensure we have a "latest" component, as well as one folder that is strictly version-pinned.
  ```yaml
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{ .Version }}"
        - "components/terraform/infra/vpc-flow-logs-bucket/latest"
  ```

```
```

- `included_paths` and `excluded_paths` support [POSIX-style greedy Globs](https://en.wikipedia.org/wiki/Glob_(programming)) for filenames/paths
  (double-star/globstar `**` is supported as well)

- The `component` attribute in each source is optional. It's used in the `atmos vendor pull -- component <component` command if the component is
  passed in. In this case, Atmos will vendor only the specified component instead of vendoring all the artifacts configured in the `vendor.yaml`
  manifest

> - The `source` and `targets` attributes support [Go templates](https://pkg.go.dev/text/template)
>   and [Sprig Functions](http://masterminds.github.io/sprig/). This can be used to templatise the `source` and `targets` paths with the artifact
>   versions specified in the `version` attribute

Reviewer 2: ```suggestion
- The `source` and `targets` attributes support [Go templates](https://pkg.go.dev/text/template)
  and [Sprig Functions](http://masterminds.github.io/sprig/). This can be used to templatize the `source` and `targets` paths with the artifact
  versions specified in the `version` attribute
 
  Here's an advanced example showcasing how templates can be used together with `targets` 
  ```yaml
      targets:
        # Vendor a component into a major-minor versioned folder like 1.2
        - "components/terraform/infra/vpc-flow-logs-bucket/{{ (first 2 (splitList \".\" .Version)) | join \".\" }}"
  ```

```

- The `imports` section defines the additional vendoring manifests that are merged into the main manifest. Hierarchical imports are supported
  at many levels (one vendoring manifest can import another, which in turn can import other manifests, etc.). Atmos processes all imports and all
  sources in the imported manifests in the order they are defined

## Hierarchical Imports in Vendoring Manifests

Use `imports` to split the main `vendor.yaml` manifest into smaller files for maintainability, or by their roles in the infrastructure.

For example, import separate manifests for networking, security, data management, CI/CD, and other layers:

```yaml
imports:
  - "layers/networking.yaml"
  - "layers/security.yaml"
  - "layers/data.yaml"
  - "layers/analytics.yaml"
  - "layers/firewalls.yaml"

>   - "layers/cicd.yaml"

Reviewer 2: Is the yaml extension required?

```

Hierarchical imports are supported at many levels. For example, consider the following vendoring configurations:

```yaml title="vendor.yaml"
apiVersion: atmos/v1
kind: AtmosVendorConfig
metadata:

>   name: atmos-vendor-config

Reviewer 2: ```suggestion
  name: example-vendor-config
```

      version: "1.323.0"
      targets:
        - "components/terraform/infra/vpc-flow-logs-bucket/{{.Version}}"
```

```yaml title="vendor/vendor2.yaml"
apiVersion: atmos/v1
kind: AtmosVendorConfig
metadata:
  name: example-vendor-config-2
  description: Atmos vendoring manifest
spec:
  imports:
    - "vendor/vendor4.yaml"

  sources:
    - component: "my-vpc1"
      source: "oci://public.ecr.aws/cloudposse/components/terraform/stable/aws/vpc:{{.Version}}"
      version: "1.0.2"
      targets:
        - "components/terraform/infra/my-vpc1"
```

```yaml title="vendor/vendor4.yaml"
apiVersion: atmos/v1
kind: AtmosVendorConfig
metadata:
  name: example-vendor-config-4
  description: Atmos vendoring manifest
spec:
  imports:
    - "vendor/vendor5.yaml"

  sources:
    - component: "my-vpc4"
      source: "github.com/cloudposse/terraform-aws-components.git//modules/vpc?ref={{.Version}}"
      version: "1.319.0"
      targets:
        - "components/terraform/infra/my-vpc4"
```

When you execute the `atmos vendor pull` command, Atmos processes the import chain and the sources in the imported manifests in the order they
are defined:

> - The `vendor/vendor2.yaml` and `vendor/vendor3.yaml` manifests (defined in the main `vendor.yaml` file) are imported
> - The `vendor/vendor2.yaml` file is processed, and the `vendor/vendor4.yaml` manifest is imported
> - The `vendor/vendor4.yaml` file is processed, and the `vendor/vendor5.yaml` manifest is imported
> - Etc.

Reviewer 2: ```suggestion
- First, the main `vendor.yaml` file is read based on search paths
- The `vendor/vendor2.yaml` and `vendor/vendor3.yaml` manifests (defined in the main `vendor.yaml` file) are imported
- The `vendor/vendor2.yaml` file is processed, and the `vendor/vendor4.yaml` manifest is imported
- The `vendor/vendor4.yaml` file is processed, and the `vendor/vendor5.yaml` manifest is imported
- Etc.
```
